cmd/soroban-rpc/internal/events/events.go

```diff
@@ -0,0 +1,276 @@
+		Event:  math.MaxUint32,
+	}
+)
+
+type bucket struct {
+	ledgerSeq uint32
+	events    []event
+}
+
+type event struct {
+	contents   xdr.ContractEvent
+	txIndex    uint32
+	opIndex    uint32
+	eventIndex uint32
+}
+
+func (e event) cursor(ledger uint32) Cursor {
+	return Cursor{
+		Ledger: ledger,
+		Tx:     e.txIndex,
+		Op:     e.opIndex,
+		Event:  e.eventIndex,
+	}
+}
+
+// MemoryStore is an in-memory store of soroban events.
+type MemoryStore struct {
+	lock sync.RWMutex
+	// buckets is a circular buffer where each cell represents
+	// all events occurring within a specific ledger.
+	buckets []bucket
+	// length is equal to the number of ledgers contained within
+	// the circular buffer.
+	length uint32
+	// start is the index of the head in the circular buffer.
+	start uint32
+}
+
+// NewMemoryStore creates a new MemoryStore.
+// The retention window is in units of ledgers.
+// All events occurring in the following ledger range
+// [ latestLedger - retentionWindow, latestLedger ]
+// will be included in the MemoryStore. If the MemoryStore
+// is full, any events from new ledgers will evict
+// older entries outside the retention window.
+func NewMemoryStore(retentionWindow uint32) (*MemoryStore, error) {
+	if retentionWindow == 0 {
+		return nil, fmt.Errorf("retention window must be positive")
+	}
+	return &MemoryStore{
+		buckets: make([]bucket, retentionWindow),
+	}, nil
+}
+
+// Range defines a [Start, End) interval of Soroban events.
+type Range struct {
+	// Start defines the (inclusive) start of the range.
+	Start Cursor
+	// ClampStart indicates whether Start should be clamped up
+	// to the earliest ledger available if Start is too low.
+	ClampStart bool
+	// End defines the (exclusive) end of the range.
+	End Cursor
+	// ClampEnd indicates whether End should be clamped down
+	// to the latest ledger available if End is too high.
+	ClampEnd bool
+}
+
+// Scan applies f on all the events occurring in the given range.
+// The events are processed in sorted ascending Cursor order.
+// If f returns false, the scan terminates early (f will not be applied on
+// remaining events in the range). Note that a read lock is held for the
+// entire duration of the Scan function so f should be written in a way
+// to minimize latency.
+func (m *MemoryStore) Scan(eventRange Range, f func(Cursor, xdr.ContractEvent) bool) error {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+
+	if err := m.validateRange(&eventRange); err != nil {
+		return err
+	}
+
+	curLedger := eventRange.Start.Ledger
+	minLedger := m.buckets[m.start].ledgerSeq
+	i := ((curLedger - minLedger) + m.start) % uint32(len(m.buckets))
+	events := seek(m.buckets[i].events, eventRange.Start)
+	for {
+		for _, event := range events {
+			cur := event.cursor(curLedger)
+			if eventRange.End.Cmp(cur) <= 0 {
+				return nil
+			}
+			if !f(cur, event.contents) {
+				return nil
+			}
+		}
+		i = (i + 1) % uint32(len(m.buckets))
+		curLedger++
+		if m.buckets[i].ledgerSeq != curLedger {
```

Comment:
this for loop isn't incorrect, but it makes me nervous since there is no obvious termination condition.
can you move the `m.buckets[i].ledgerSeq != curLedger` to be the terminating condition for the loop ?
i.e.
```golang
for ;m.buckets[i].ledgerSeq == curLedger; curLedger++ {
...
}
```